src/msw/dialog.cpp

```diff
@@ -181,17 +181,15 @@ int wxDialog::ShowModal()
 
     wxASSERT_MSG( !IsModal(), wxT("ShowModal() can't be called twice") );
 
-    Show();
+    //enter and run the modal loop
+    wxDialogModalDataTiedPtr modalData(&m_modalData,
+                                       new wxDialogModalData(this));
 
+    Show();
     // EndModal may have been called from InitDialog handler (called from
     // inside Show()) and hidden the dialog back again
-    if ( IsShown() )
```

Comment:
Sorry, I don't understand this. I think it breaks the use case mentioned in the comment (which, BTW, doesn't make sense any more if this test is removed), i.e. what happens if `wxEVT_INIT_DIALOG` handler calls `EndModel()`? It's supposed to result in not showing the dialog at all, but it seems that with this change it will show it and run the modal loop.

Please check this interactively and, ideally, also add the unit test to check for this too.